writeaheadlog_integration_test.go

```diff
@@ -1,556 +1,487 @@
-	prefix := make([]byte, 8)
-	binary.LittleEndian.PutUint64(prefix, csu.index)
-	return append(prefix, csu.splotchData...)
-}
+		// number is the number which is written at offset
+		number uint32
 
-// unmarshalChangeSplotchUpdate will unpack a set of updates.
-func unmarshalChangeSplotchUpdate(updateBytes []byte) changeSplotchUpdate {
-	return changeSplotchUpdate{
-		index:       binary.LittleEndian.Uint64(updateBytes[:8]),
-		splotchData: updateBytes[8:],
-	}
-}
+		// silo is the offset of the silo so we can apply the update to the right silo
+		silo int64
 
-// countingArray is a sample object that we wish to persist to disk using the
-// wal.
-type countdownArray struct {
-	// Should always be {2, 1, 0}, or {3, 2, 1, 0}... etc. The file contains an
-	// array on disk that should have the exact same property.
-	countdown []uint64
-	file      *os.File
-
-	// splotchFile is an extension to the countdown file, where large pages of
-	// the file are all required to have a continuous countdown. The count can
-	// start at any value, and then it will count down for the whole page,
-	// underflowing if required.
-	//
-	// The first 80 bytes must be 10 8-byte values that count down and underflow.
-	// The next 160 bytes must be 20 8-byte values that count down and underflow.
-	// The next 240 bytes must be 30 8-byte values that count down and underflow.
-	// ...
-	//
-	// Different parts of this file can easily be edited in parallel
-	// transactions, while at the same time there are clear consistency
-	// requirements.
-	//
-	// '0' is used as a special starting value to indicate that the whole
-	// splotch should be empty.
-	splotchFile *os.File
-
-	wal *WAL
-}
+		// ocs is the checksum of the file that became obsolete after changing a
+		// number. It needs to be removed when the update is applied.
+		ocs checksum
 
-// addCount will increment every counter in the array, and then append a '1',
-// but using the wal to maintain consistency on disk.
-func (ca *countdownArray) addCount() error {
-	// Increment the count in memory, creating a list of updates as we go.
-	var updates []Update
-	for i := 0; i < len(ca.countdown); i++ {
-		ca.countdown[i]++
-		updates = append(updates, Update{
-			Name:    addCountName,
-			Version: "1.0.0",
-			Instructions: addCountUpdate{
-				index:    uint64(i),
-				newCount: ca.countdown[i],
-			}.marshal(),
-		})
-	}
+		// ncs is the new checkum of the silo and the corresponding file
+		ncs checksum
 
-	// Corner case - if there are no updates, because this is the first time we
-	// add to the count, just write out the file piece.
-	if len(updates) == 0 {
-		writeBytes := make([]byte, 8)
-		_, err := ca.file.WriteAt(writeBytes, 8*int64(len(ca.countdown)))
-		if err != nil {
-			return err
-		}
-		err = ca.file.Sync()
-		if err != nil {
-			return err
-		}
-		ca.countdown = append(ca.countdown, 0)
-		return nil
+		// ncso is the offset of the new checksum
+		ncso int64
```

Comment:
It's especially important to spell out the names of variables within structs, with very few exceptions, because you are going to reference the structs out-of-context a bunch of times, and it's easy for people unfamiliar with the code to get lost.

Reply:
ok